### ISSUE TITLE
Linter plugins: Pass globals to JS plugins

### ISSUE THREAD

**Issue by overlookmotel** [oxc-project/oxc#16359]
Globals can be specified in Oxlint config files, and are already parsed on Rust side.

We need to send them over to JS side, and make them available to JS plugins as `context.languageOptions.globals`.

I think we can use the same technique as #16215 used for passing options from Rust to JS - send them over once after parsing all configs, and reference them with IDs thereafter.

**Comment by lilnasy** [comment]
For background, #16215 introduced `setup_configs()`, which runs once per process and gives oxlint an opportunity to serialize data that does **not** change between files to lint. Rule options worked easily for this approach because they are not subject to merging and overriding so they can be marshaled once then used as-is.

I think globals, and settings for that matter, will not be as simple -- at least, it will not be simple if we want to support merging multiple configurations (actually, I am not sure if merging is supported even for internal plugins yet, but it's a consideration regardless) and overrides, supported since #8915.
https://github.com/oxc-project/oxc/blob/8c85e087f0e9f087988e145ad1e9bd4afbdbf066/crates/oxc_linter/src/config/config_store.rs#L235-L237

E2E test for this current behavior: [config](https://github.com/oxc-project/oxc/pull/16512/files#diff-a56a021c38ec859aa879b884a91e60d68386dc2383139ac6881d2fc2aa42cb2a), [snapshot](https://github.com/oxc-project/oxc/pull/16512/files#diff-041648a519d0a66cdb7eeff68ff64f9c5135c11a8ef89e2e9557e93ba24fe1a6). The test is passing with straightforward (but inefficient) serialization of globals for every file: [oxc_linter/src/lib.rs](https://github.com/oxc-project/oxc/pull/16512/files#diff-207225884c5e031ffd802bb99e4fbacbd8364b1343a1cec5485bf50f29186300).

## Heavyweight approach
My first idea was to serialize the base globals and override globals as-is, and reimplement some of the resolve, glob-match, override logic on the JS side. I have gotten it working with the tests passing, but I am not happy with it. I don't like the fact that it required an additional dependency (analog to fast-glob on the rust side), and that it would likely snowball into more rust logic being rewritten in JS, which is is the opposite of what we generally want. We need bit of a smarter solution.

## Smart approach
We perform the overrides ahead of time on the rust side, and correlate the merged/overriden sets of globals with the files they apply to -- also ahead of time since we should have access to the full list of files to lint. We can serialize these somewhat interrelated sets once in `setup_configs` as an array, then serialize the index of the effective set of globals in `lint_file()`.

This is all very similar to how rule options are handled, but with more preprocessing. There will be some duplication in the array of globals, but the memory impact is probably negligible as very few projects will have more than single digit effective sets of globals anyway. This is what I will try next.

**Comment by overlookmotel** [comment]
As I mentioned on #16512, I think we should merge that PR as a starting point.

I thought globals weren't merged between main config and overrides - thought the override just "wins". But judging from the test case in #16512, I was wrong about that.

I agree with your "smart approach". But... I think merging is going to complicate things at lot.

My *guess* is that merging happens per file during linting - not during config parsing - because any combination of different overrides could apply to a file, and you don't know which combinations are going to be used in advance. You don't want to pre-calculate the merged state of every possible combination of overrides in advance (at config-parsing time) because if there are e.g. 20 overrides, the number of possible combinations would be enormous.

So we'd need to store some kind of mapping from `Vec<OverrideId>` -> `Globals`, to assign each merged globals object an ID when it's first created, and then reuse that same ID for the next time that combination of overrides is used again.

Linting happens on multiple threads, so the mapping will need to be thread-safe (wrapped in a `Mutex` or something). *And* we'd need a way to "stream" new globals objects over to JS during linting. Ergh!

So, in short, not easy... consequently, I'd be keen to merge the "inefficient but works" version in #16512 first.

**Comment by overlookmotel** [comment]
IMO it'd probably be better to make getting settings from Rust to JS more efficient first. I *believe* that settings are not merged, in which case it'll be simpler. But doing that we'll start to get an idea of good design patterns. Then we work up to dealing with globals.

**Comment by lilnasy** [comment]
> My guess is that merging happens per file during linting 

Yes, that's correct.

> because any combination of different overrides could apply to a file, and you don't know which combinations are going to be used in advance

In my feature branch, I was able to get far enough that we know which combinations are going to be used in advance. It just resulted in duplicated code that I haven't thought about how to properly refactor. 

> You don't want to pre-calculate the merged state of every possible combination of overrides in advance (at config-parsing time) because if there are e.g. 20 overrides, the number of possible combinations would be enormous.

That's true, but how common of a case is that?

**Comment by overlookmotel** [comment]
> In my feature branch, I was able to get far enough that we know which combinations are going to be used in advance. It just resulted in duplicated code that I haven't thought about how to properly refactor.

Interesting. How did you manage that?

> > You don't want to pre-calculate the merged state of every possible combination of overrides in advance (at config-parsing time) because if there are e.g. 20 overrides, the number of possible combinations would be enormous.
> 
> That's true, but how common of a case is that?

Not common I'd imagine, but also not completely out of the question. ESLint's flat config style encourages you to define a bunch of configs in one array for e.g. `packages/a/**/*.ts`, `packages/b/**/*.ts`, `packages/a/tests/**/*.ts` etc. And overrides is our equivalent of that pattern.

Anyway I've learned from bitter experience that every time you think "nobody would do that", someone does it.

If it was an uncommon case *and* only slight perf cost, then we could probably go with it. But in this case the perf impact is exponential over number of overrides, so the worst case scenario is really bad.

**Comment by overlookmotel** [comment]
Implemented in #16512. Leaving this issue open for improving the efficiency of the implementation.

Also:

* [ ] Support `env` config (or add tests if we do already) - see https://github.com/oxc-project/oxc/pull/16512#pullrequestreview-3547756955
* [x] Rename `GlobalValue::Writeable` to `Writable` - #16609 + #16610.

On the last one: We could just add a `#[serde(rename = "writable")]` attr to the enum variant to fix it in the serialized JSON. But since `Writeable` is a mis-spelling, I think it's preferable to rename it fully in the Rust type. Thankfully, Rust Analyser makes that easy.

**Comment by connorshea** [comment]
Are we currently [supporting true/false values for globals](https://eslint.org/docs/latest/use/configure/language-options#using-configuration-files)? They're technically allowed, but I'm unsure if they're being interpreted/converted correctly before being sent to the jsPlugin right now. It seems like a lot of the remaining errors in the conformance tests may involve that.

e.g.

```
#### no-implied-eval > invalid

\```js
setTimeout("x = 1;");
\```
```

Where `setTimeout` is a global set with `false` in the given test.

EDIT: I added a test for this in https://github.com/oxc-project/oxc/pull/16860 and it seems to work fine with a boolean value, so I'm probably wrong about this assumption.

**Comment by overlookmotel** [comment]
We currently don't pass globals into rules at all in `RuleTester`!

https://github.com/oxc-project/oxc/blob/faca352690939954cd4bc9036309b3d090ce00ef/apps/oxlint/src-js/package/rule_tester.ts#L965-L967

We do in real Oxlint, though, and Rust-side code should correctly handle `true` and `false`. Oh actually, your recent PR proves that works.

**Comment by overlookmotel** [comment]
Globals are now passed correctly in `RuleTester` (#17009).
